inc/namespace.php

```diff
@@ -66,6 +74,35 @@ function on_shutdown() {
 	send_trace_to_daemon( get_end_trace() );
 }
 
+/**
+ * Shutdown callback that is triggered via register_shutdown_function.
+ *
+ * In some cases the on_shutdown_action() function will not be called, so we have
+ * this failsafe shutdown function to catch any cases where on_shutdown_action() is not called.
+ *
+ */
+function on_shutdown() {
+	// If we shutdown before the plugin API has loaded, return early.
+	if ( ! function_exists( 'has_action' ) ) {
+		on_shutdown_action();
+		return;
+	}
+	// If the "shutdown_action_hook" function has not been registered
+	// to the shutdown action, call it now.
+	if ( has_action( 'shutdown', __NAMESPACE__ . '\\on_shutdown_action' ) === false ) {
+		on_shutdown_action();
+		return;
+	}
+
+	// It's possible the script is shutting down before register_shutdown_function( 'shutdown_action_hook' )
+	// has been called by WordPress. There's no way to check if this callback has been registered, so we use a heuristic that
+	// default-filters.php has not been loaded, which happens just before WordPress calls register_shutdown_function.
+	if ( has_action( 'wp_scheduled_delete', 'wp_scheduled_delete' ) === false ) {
```

Comment:
`function_exists( 'get_locale' )` would be a better heuristic; `l10n.php` is loaded immediately after the shutdown function.

As an alternative idea, could we call `register_shutdown_function( 'shutdown_action_hook' )` ourselves? What happens when the same function is added twice?

Reply:
Yeah I didn't want to do that, as I think there might be a look of functionality hooked into `shutdown` that cached-responses might not want.